### ISSUE TITLE
Move over to using `/run/motd.d/` instead of `/run/console-login-helper-messages/motd.d/` for the default included information generators units

### ISSUE THREAD

**Issue by rfairley** [coreos/console-login-helper-messages#60]
#57 will introduce separate units such as `console-login-helper-messages-gensippet-os-release.service` for sourcing information and writing out a MOTD snippet including the information, to be displayed at SSH login. The script this service runs, `gensnippet_os_release`, specifies an output directory, which is not exposed outside of the script. After some time, once [necessary pieces in PAM and SELinux](https://src.fedoraproject.org/rpms/console-login-helper-messages/blob/master/f/console-login-helper-messages.spec#_52) land in RHEL, services like this should be switched to just write a snippet within `/run/motd.d` rather than `/run/console-login-helper-messages/motd.d/`.

Alternatively, make the output directory configurable (e.g. through an `Environment=` var through the service unit). In Fedora 32+, all the needed bits in external packages are in place to use `/run/motd.d` (as root user). However, there isn't any major issue with going through `/run/console-login-helper-messages/motd.d/` for now - waiting is still fine.

(Additionally once the upstream bits land in RHEL, the RHCOS manifest should be updated include `console-login-helper-messages-motdgen` like FCOS does.)

Related: #7

**Comment by kelvinfan001** [comment]
With `util-linux 2.36` (which will be in Fedora 33), there is now support to let `login` read from custom directories to display MOTDs after login through the serial console. The default directories that it reads from are `/usr/share/misc/motd:/run/motd:/etc/motd` ([man page](https://man7.org/linux/man-pages/man1/login.1.html)). It is possible to add, e.g. `/run/motd.d` to `/etc/login.defs`'s `MOTD_FILE=` item to let `login` display MOTD snippets from `/run/motd.d`.
However, `/etc/login.defs` is owned by `shadow-utils`, and `login.defs` doesn't seem to support dropin directories. Are there other options for CLHM to configure `login.defs`? Based on [this commit](https://github.com/karelzak/util-linux/commit/72b155ea6e25730d7c01d345cc2df269c2c47635) from `util-linux`, it seems unlikely that we can add `/run/motd.d` as a default for `login`. What do you think? @jlebon @dustymabe

**Comment by dustymabe** [comment]
@jlebon - any thoughts here? my understanding is that if we edit `login.defs` during compose then we own that file, which might not be what we want. I wish it had a dropin config structure like `/etc/login.defs.d/` where we could configure things that way.

**Comment by jlebon** [comment]
> my understanding is that if we edit `login.defs` during compose then we own that file, which might not be what we want

Hmm, can you expand here? Modifications at compose time would still be part of the OSTree commit, whereas changing it on the client-side means it becomes untracked. So I think it's fine for this to just modify it at compose time (@kelvinfan001 check out the `postprocess` scripts in the f-c-c repo for examples).

**Comment by kelvinfan001** [comment]
@jlebon You mean something like what's in here https://github.com/coreos/fedora-coreos-config/blob/2d898061f72af61e6dac9849ddf0545ed8b9a3ed/manifests/fedora-coreos-base.yaml#L65 (i.e. we use `postprocess` scripts to modify the contents of `/etc/login.defs`, and it'll still be in the FCOS image)? Can I do the same to modify how `agetty` is called by systemd (by modifying the contents of `/lib/systemd/system/getty@.service`)?

**Comment by dustymabe** [comment]
What should we do on non FCOS systems? We're using CLHM in the Fedora Cloud base image too.

**Comment by jlebon** [comment]
Pinging @karelzak and @ikerexxe, maintainers of util-linux and shadow-utils in Fedora: we'd like to have `login` support reading MOTD files from `/run/motd.d`. Do I understand correctly from the commit message in https://github.com/karelzak/util-linux/commit/72b155ea6e25730d7c01d345cc2df269c2c47635 that you'd be open to doing this at the distro level by changing the default fallback for `MOTD_FILE`  for `login`?

Alternatively, we can make this a Fedora CoreOS thing, but the current structure doesn't make it easy. Ideally, we would also have an `/etc/login.defs.d` directory and then we could ship a drop-in there in Fedora CoreOS to set `MOTD_FILE`. How complex would that be?

In the worst case, we can modify `/etc/login.defs` directly, though it's not as elegant.

**Comment by ikerexxe** [comment]
> Pinging @karelzak and @ikerexxe, maintainers of util-linux and shadow-utils in Fedora: we'd like to have `login` support reading MOTD files from `/run/motd.d`. Do I understand correctly from the commit message in [karelzak/util-linux@72b155e](https://github.com/karelzak/util-linux/commit/72b155ea6e25730d7c01d345cc2df269c2c47635) that you'd be open to doing this at the distro level by changing the default fallback for `MOTD_FILE` for `login`?
> 

AFAIK, `MOTD_FILE` is not currently supported in Fedora. I think it's related to how login is implemented in that distro, but not completely sure. Maybe @karelzak can shed some light.

> Alternatively, we can make this a Fedora CoreOS thing, but the current structure doesn't make it easy. Ideally, we would also have an `/etc/login.defs.d` directory and then we could ship a drop-in there in Fedora CoreOS to set `MOTD_FILE`. How complex would that be?
> 

Definitely not a good idea. This would be a RFE for shadow-utils that should be discussed upstream and I don't know how it will be received. What I know for sure is that it will take some time to get accepted.

> In the worst case, we can modify `/etc/login.defs` directly, though it's not as elegant.

I wouldn't mind changing this file in shadow-utils dist-git as long as it is a simple change. Though, if the change is related with `MOTD_FILE` then we come again to the first part of my comment: that currently it isn't supported.

**Comment by karelzak** [comment]
MOTD_FILE is supported by util-linux login(1) since v2.36, and it's up to the distribution maintainers to set the hardcoded default (during package compilation) or to modify distributed /etc/login.defs. 

For Fedora, I have no problem adding /run/motd.d to the hardcoded MOTH_FILE. It's a trivial patch we can maintain in util-linux rpm.

ad /etc/login.defs.d -- I'm not sure about it. It's a very basic configuration, and not sure if change it by package installation is got idea. (BTW, SUSE uses libeconf as a replacement to login.defs. It supports all the crazy things, overwriting config options, multiple files, etc.)

IMHO the best is to keep it easy -- use /run/motd.d in distros and later change the default also in upstream.

**Comment by kelvinfan001** [comment]
> use /run/motd.d in distros and later change the default also in upstream 

@karelzak Thanks. This way sounds good to me! I've opened a PR to apply the patch https://src.fedoraproject.org/rpms/util-linux/pull-request/7.

**Comment by kelvinfan001** [comment]
@karelzak Similar to our issues with `login(1)` that @jlebon mentioned here https://github.com/coreos/console-login-helper-messages/issues/60#issuecomment-737501944. It would be nice if `/run/issue.d` were a directory that `agetty` read from by default. Would you be open to doing this at the distro level by changing the default for `--issue-file` for `agetty`?

**Comment by karelzak** [comment]
> @karelzak Similar to our issues with `login(1)` that @jlebon mentioned here [#60 (comment)](https://github.com/coreos/console-login-helper-messages/issues/60#issuecomment-737501944). It would be nice if `/run/issue.d` were a directory that `agetty` read from by default. Would you be open to doing this at the distro level by changing the default for `--issue-file` for `agetty`?

The issue file in agetty has a little bit different semantic (due to backward compatibility) than motd. 

agetty(8) reads from /run by default if you do NOT have /etc/issue, the command line option `--issue-file` is a way how to ignore /etc (and other default locations). The option makes sense for admins who do not want to touch distro default, but they need something else or a terminal/line specific setting. 

It does not make sense to use the option as the distribution default, it's better to not use /etc/issue at all -- or if you want to use /run for some subset of the distribution (virtual machines) than use `mv /etc/issue /etc/issue.unused` in your generator or package install script.

man agetty:
```
ISSUE FILES
       The default issue file is /etc/issue. If the file exists, then agetty also checks for /etc/issue.d directory. The  di‐
       rectory  is optional extension to the default issue file and content of the directory is printed after /etc/issue con‐
       tent. If the /etc/issue does not exist, then the directory is ignored. All files with .issue extension from the direc‐
       tory  are printed in version-sort order. The directory can be used to maintain 3rd-party messages independently on the
       primary system /etc/issue file.

       Since version 2.35 additional locations for issue file and directory are supported. If the default /etc/issue does not
       exist,  then  agetty  checks for /run/issue and /run/issue.d, thereafter for /usr/lib/issue and /usr/lib/issue.d.  The
       directory /etc is expected for host specific configuration, /run is expected for  generated  stuff  and  /usr/lib  for
       static distribution maintained configuration.

       The  default path maybe overridden by --issue-file option. In this case specified path has to be file or directory and
       all the default issue file and directory locations are ignored.
```

**Comment by kelvinfan001** [comment]
@karelzak Thank you for the explanation!

**Comment by kelvinfan001** [comment]
Closed by #91

**Comment by karelzak** [comment]
> > use /run/motd.d in distros and later change the default also in upstream
> 
> @karelzak Thanks. This way sounds good to me! I've opened a PR to apply the patch https://src.fedoraproject.org/rpms/util-linux/pull-request/7.

@kelvinfan001  the change should be available in util-linux-2.36.1-4.fc34. (sorry for the delay ...)
